### DISCUSSION TITLE
Deploy glance on Vercel or Netlify or Railway

### DISCUSSION THREAD

**Discussion by vishalkadam47** [glanceapp/glance#185] [Help] [answered]
is there a way to Deploy glance on Vercel or Netlify, any config required i did try with netlify but facing some go version errors and if I change the version as I get dependencies error.
if anyone has working config vercel.json or netlify.toml would be helpful 
or render.com

**Comment by ken-morel** [comment]
I am not sure vercel supports golang... but one option I have used was, railway.
pros:
- Supports several build languages
- Provide interesting project structure
- Beautiful UI
cons:
- it is Paid 😞, offers $5 free trial, if you just deploy glance instance, should take months to finish(exept you need to frequently update glance.yml)

I dropped my glance instance, but deploying it just now for you to have an example...

**Reply by vishalkadam47** [reply]
i found this doc but don't know how to configure https://vercel.com/docs/functions/runtimes/go

**Reply by vishalkadam47** [reply]
thanks for the dockerfile, just logged in to railway, I selected the repo and deployment was successful but how do i add, edit or update glance.yml
```
failed opening config file: open glance.yml: no such file or directory

failed opening config file: open glance.yml: no such file or directory

container event container restart

container event container died

```

**Reply by ken-morel** [reply]
Sorry, let me try try find back the fix I used last time(I am really poor in Dockerfile). The dockerfile I provided fails to copy the glance.yml into the path where glance will find it. You could try to copy the glance.yml or modify where glance searches for it.

**Reply by vishalkadam47** [reply]
do i need to add `glance.yml` file in the repo

**Reply by ken-morel** [reply]
Yes, it will take its configuration from there. But the dockerfile should properly copy the glance.yml

**Reply by ken-morel** [reply]
My glance instance at https://glance-production.up.railway.app/
I could not use normal docker to copy the glance.yml to the root directory, it seems the repository is not entirely cloned... but a turnaround was to download the glance.yml using curl, here my code

```Dockerfile  
FROM golang:1.22-alpine AS builder  
WORKDIR /app  
COPY go.mod go.sum ./
RUN go mod download  
COPY . .  
RUN go build -o myapp .  
FROM alpine:latest
# Install curl to download glance.yml from repository
RUN apk add --no-cache curl  
WORKDIR /app  
# update the username and repository names to yours please
RUN curl -o glance.yml https://raw.githubusercontent.com/ken-morel/glance/main/glance.yml    
COPY --from=builder /app/myapp /myapp  
ENTRYPOINT ["/myapp"]
```

**Reply by vishalkadam47** [reply]
I'm using https://gist.github.com/ for `glance.yml`, now I can edit and update `yml` whenever required without unwanted commits

**Comment by ken-morel** [comment]
My glance instance at https://glance-production.up.railway.app/
I could not use normal docker to copy the glance.yml to the root directory, it seems the repository is not entirely cloned... but a turnaround was to download the glance.yml using curl, here my code

```Dockerfile  
FROM golang:1.22-alpine AS builder  
WORKDIR /app  
COPY go.mod go.sum ./
RUN go mod download  
COPY . .  
RUN go build -o myapp .  
FROM alpine:latest
# Install curl to download glance.yml from repository
RUN apk add --no-cache curl  
WORKDIR /app  
# update the username and repository names to yours please
RUN curl -o glance.yml https://raw.githubusercontent.com/ken-morel/glance/main/glance.yml    
COPY --from=builder /app/myapp /myapp  
ENTRYPOINT ["/myapp"]
```

**Reply by vishalkadam47** [reply]
thanks, it worked but only reddit links are showing error
![image](https://github.com/user-attachments/assets/426fab31-518c-4578-910c-710b90af8a16)

**Reply by ken-morel** [reply]
Sorry, surely does reddit not support requests from railway services public IPs, I donot know a fix for that

**Reply by vishalkadam47** [reply]
no problem for now I have removed reddit links

**Reply by svilenmarkov** [reply]
This is a known limitation - Reddit blocks VPS IPs. You can get around this by routing the traffic from Glance through a VPN or routing through a proxy hosted on your home network and using the `request-url-template` property.

**Comment by ken-morel** [comment]


**Reply by ken-morel** [reply]
I tried moving the index.go to repository root, but now vercel does not detect it and serves the page as static. The only way seems to adapt glance to the `/api` based urls from vercel... or to build other packages for other routes... working on that!

**Reply by ken-morel** [reply]
Well, seems vercel can definitely not deploy glance:
- It only permits index.go file in `/api/` from my seen
what does not adapt to glance's routing, I am working on a special thing which will adapt to that. But you will need to do some changes

**Reply by ken-morel** [reply]
well. gave up!

**Reply by svilenmarkov** [reply]
I'm not sure this is worth the hassle as Glance was never meant to work in a serverless environment. The whole point of serverless is you accept a request, run a function, respond and then shut down. Glance relies on being a long running process that can cache and reuse widget data in memory. You'd have to rewrite a lot of stuff to get it to play nicely with serverless. Not impossible, just... not worth it when you can run a dozen instances of Glance on a $15 Raspberry Pi Zero W.

**Reply by vishalkadam47** [reply]
@ken-morel thanks buddy for taking time and thanks for helping me with `railway`.
I was just looking for a way to deploy it somewhere in the cloud publicly and https://railway.app/ works as I was looking for, now I don't have to pay for a VM server just to run `Glanceapp`
@svilenmarkov Thanks for the information, I have used vercel and netlify for other dashboard apps so I thought Glance would also work.

**Comment by vishalkadam47** [comment]
@ken-morel 
when using `custom.css` do we need to add any line of code in `Dockerfile` because as of now if i use `assets-path` in `glance.yml` Is not working when redeployed. please guide me
```
server:
  assets-path: ./assets
theme:
  custom-css-file: "/assets/custom.css"
```

**Reply by ken-morel** [reply]
Oh, :laughing: I feel so stupid. I forgot the `.dockerignore`!
Please, simply delete the `.dockerignore`, and modify the dockerfile to something more simple:

```dockerfile
FROM golang:1.22-alpine AS builder
COPY . .
RUN go mod download
RUN go build -o /glance .
ENTRYPOINT ["/glance"]
```

This worked for me.
My deployment at https://glance.up.railway.app/

**Reply by ken-morel** [reply]
Please, @vishalkadam47 You selected the answer earlier, update it to that just above.

**Reply by vishalkadam47** [reply]
deployment works but the css is not working

**Reply by ken-morel** [reply]
No really!, mine has no issue, please compare your repository to mine at https://github.com/ken-morel/glance (notice the `assets/custom.css` and deleted `.dockerignore`) and view the output at https://glance.up.railway.app/

**Comment by ken-morel** [comment, marked as answer]
Delete the `.dockerignore`, and modify the `.dockerfile`:

```dockerfile
FROM golang:1.22-alpine AS builder
COPY . .
RUN go mod download
RUN go build -o /glance .
ENTRYPOINT ["/glance"]
```

This worked for me.
My deployment at https://glance.up.railway.app/ (including custom css)

This implies all your assets including glance.yml are present in the repository.

**Reply by vishalkadam47** [reply]
got it working now, I missed the part of deleting `.dockerignore`

**Comment by vishalkadam47** [comment]
I'm using glance.yml from Gist 
```Dockerfile
FROM golang:1.22.5-alpine
WORKDIR /app
COPY . .
RUN go mod download
RUN go build -o app .
# Install curl to download glance.yml
RUN apk add --no-cache curl
# Download glance.yml from the specified gist
ARG GLANCE_FILE
RUN curl -o glance.yml $GLANCE_FILE
EXPOSE 8080
ENTRYPOINT ["./app"]
```

**Comment by vishalkadam47** [comment]
Deploy Glance on [railway.com](https://railway.com/template/ZHBnkG?referralCode=vishify)

- added environment variable to downnload `glance.yml`
> GLANCE_CONFIG_SOURCE=https://raw.githubusercontent.com/glanceapp/glance/refs/heads/main/docs/glance.yml

[![Deploy on Railway](https://railway.com/button.svg)](https://railway.com/template/ZHBnkG?referralCode=vishify)

- `start command` which i have used in the  template to download the `glance.yml` using env , but i guess this can be enhanced more to support multiple `yml` files
```
/bin/sh -c 'if [ ! -f /app/config/glance.yml ]; then echo "Config not found in volume, downloading configuration..." && wget -O /app/config/glance.yml $GLANCE_CONFIG_SOURCE && echo "Configuration has been populated in /app/config/glance.yml"; else echo "Configuration found in volume, continuing with startup..."; fi && echo "Starting Glance..." && ./glance --config /app/config/glance.yml'
```

**Reply by vishalkadam47** [reply]
## A Clone of Glance with GitHub Gist API and single volume support
`Here's my version of glance if anyone required give it a try`

Deploy Glance on [railway.com](https://railway.com/template/UGcNz8?referralCode=vishify)

How it works
> `GIST_ID`
> - Set this to the ID of a public/private GitHub Gist that contains your text-based configuration files (such as .yml, .css, .js, etc.).
> - The app will automatically download all files in /app/config.
> 
> `CONFIG_FILES`
> - Use this as an alternative to `GIST_ID` when you want to provide direct URLs to your config files.
> - Provide a comma-separated list of URLs; each file will be downloaded and stored in /app/config.
> 
> If `GIST_ID` is set and non-empty:
> - Only the files from the Gist will be downloaded.
> - The value in `CONFIG_FILES` will be completely ignored — no files from it will be downloaded.
> This ensures there's no duplication or confusion about the source of config files.

[![Deploy on Railway](https://railway.com/button.svg)](https://railway.com/template/UGcNz8)
